processor/src/advice/mod.rs

```diff
@@ -7,188 +7,113 @@ use vm_core::{
+// need for the advice provider (or any other external component) to keep track of it. If we
+// delegate this control to the advice provider - especially if it is a trait, the user might
+// implement it incorrectly, creating undefined behavior on the VM side (that expects the counter
+// to incrementally increase).
 
+/// Common behavior of advice providers for program execution.
+///
+/// An advice provider supplies non-deterministic inputs to the processor.
+///
+/// 1. Provide a tape functionality that yields elements as a stack (last in, first out). These can
+///    be yielded as elements, words or double words.
+/// 2. Provide a map functionality that will store temporary tapes that can be appended to the main
+///    tape. This operation should not allow key overwrite; that is: if a given key exists, the
+///    implementation should error if the user attempts to insert this key again, instead of the
+///    common behavior of the maps to simply override the previous contents. This is a design
+///    decision to increase the runtime robustness of the execution.
+/// 3. Provide advice sets, that are mappings from a Merkle root its tree. The tree should yield
+///    nodes & leaves, and will provide a Merkle path if a leaf is updated.
+pub trait AdviceProvider {
     // ADVICE TAPE
     // --------------------------------------------------------------------------------------------
 
-    /// Removes the next element from the advice tape and returns it.
+    /// Pop an element from the advice tape and returns it.
     ///
     /// # Errors
     /// Returns an error if the advice tape is empty.
-    pub fn read_tape(&mut self) -> Result<Felt, ExecutionError> {
-        self.tape.pop().ok_or(ExecutionError::AdviceTapeReadFailed(self.step))
-    }
+    fn read_tape(&mut self) -> Result<Felt, ExecutionError>;
 
-    /// Removes a word (4 elements) from the advice tape and returns it.
+    /// Pops a word (4 elements) from the advice tape and returns it.
+    ///
+    /// Note: a word is always stored as little-endian. A `[...,a,b,c,d]` tape will yield
+    /// `[d,c,b,a]`.
     ///
     /// # Errors
     /// Returns an error if the advice tape does not contain a full word.
-    pub fn read_tapew(&mut self) -> Result<Word, ExecutionError> {
-        if self.tape.len() < 4 {
-            return Err(ExecutionError::AdviceTapeReadFailed(self.step));
-        }
-
-        let idx = self.tape.len() - 4;
-        let result = [self.tape[idx + 3], self.tape[idx + 2], self.tape[idx + 1], self.tape[idx]];
-
-        self.tape.truncate(idx);
-
-        Ok(result)
-    }
+    fn read_tape_w(&mut self) -> Result<Word, ExecutionError>;
 
-    /// Removes the next two words from the advice tape and returns them.
+    /// Pops a double word (8 elements) from the advice tape and returns them.
+    ///
+    /// Note: a double word is always stored as little-endian. A `[...,a,b,c,d,e,f,g,h]` tape will
+    /// yield `[h,g,f,e,d,c,b,a]`.
```

Comment:
The example result should probably be updated to represent two words: `[h, g, f, e], [d, c, b, a]`.

Reply:
Makes sense. Done!